File: hw2/triangle.cpp
//
// Created by TOPAPEC on 11.10.2021.
//

#include <cstdlib>
#include <math.h>
#include "triangle.h"

void triangle::in(FILE *fptr) {
    fscanf(fptr, "%d%d%d%d%d%d", &x1, &y1, &x2, &y2, &x3, &y3);
}

double triangle::get_side1() const {
    return sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2));
}

double triangle::get_side2() const {
    return sqrt((x1 - x3) * (x1 - x3) + (y1 - y3) * (y1 - y3));
}

double triangle::get_side3() const {
    return sqrt((x3 - x2) * (x3 - x2) + (y3 - y2) * (y3 - y2));
}

bool triangle::check_if_triangle_correct() const {
    return (get_side1() + get_side2() > get_side3()) &&
        (get_side1() + get_side3() > get_side2()) &&
        (get_side3() + get_side2() > get_side1());
}

void triangle::in_rnd() {
    x1 = rand() % 20 + 1;
    y1 = rand() % 20 + 1;

    do {
        x2 = rand() % 20 + 1;
        y2 = rand() % 20 + 1;
    } while (x2 == x1 && y1 == y2);

    do {
        x3 = rand() % 20 + 1;
        y3 = rand() % 20 + 1;
    } while (!check_if_triangle_correct());
}

void triangle::output(FILE *fptr) {
    fprintf(fptr, "It is Triangle: x1 = %d, y1 = %d, x2 = %d, y2 = %d, x3 = %d, y3 = %d. Perimeter = %f. Color = %s\n",
            x1, y1, x2, y2, x3, y3, perimeter(), shape::colors_to_string(color));
}

double triangle::perimeter() {
    return double(get_side1() + get_side2() + get_side3());
}

File: hw3/Triangle.py
import random

from Figure import Figure


class Triangle(Figure):
    def __init__(self):
        super().__init__()
        self.x1, self.y1, self.x2, self.y2, self.x3, self.y3 = 0, 0, 0, 0, 0, 0

    def input(self, args: list[str]):
        self.x1, self.y1, self.x2, self.y2, self.x3, self.y3 = \
            int(args[0]), int(args[1]), int(args[2]), int(args[3]), int(args[4]), int(args[5])

    def get_side1(self):
        return ((self.x1 - self.x2) * (self.x1 - self.x2) + (self.y1 - self.y2) * (self.y1 - self.y2)) ** (1 / 2)

    def get_side2(self):
        return ((self.x1 - self.x3) * (self.x1 - self.x3) + (self.y1 - self.y3) * (self.y1 - self.y3)) ** (1 / 2)

    def get_side3(self):
        return ((self.x3 - self.x2) * (self.x3 - self.x2) + (self.y3 - self.y2) * (self.y3 - self.y2)) ** (1 / 2)

    def check_if_triangle_is_correct(self):
        return (self.get_side1() + self.get_side2() > self.get_side3()) and \
               (self.get_side1() + self.get_side3() > self.get_side2()) and \
               (self.get_side3() + self.get_side2() > self.get_side1())

    def input_random(self):
        self.x1, self.y1 = random.randint(1, 20), random.randint(1, 20)
        self.x2, self.y2 = random.randint(1, 20), random.randint(1, 20)
        while (self.x1, self.y1) == (self.x2, self.y2):
            self.x2, self.y2 = random.randint(1, 20), random.randint(1, 20)
        self.x3, self.y3 = random.randint(1, 20), random.randint(1, 20)
        while not self.check_if_triangle_is_correct():
            self.x3, self.y3 = random.randint(1, 20), random.randint(1, 20)

    def output(self, fptr) -> None:
        fptr.write(f'It is Triangle: x1 = {self.x1}, y1 = {self.y1}, x2 = {self.x2}, y2 = {self.y2}, x3 = {self.x3},'
                   f' y3 = {self.y3}. Perimeter = {self.perimeter()}. Color = {self.color}\n')

    def print(self) -> str:
        return (f'It is Triangle: x1 = {self.x1}, y1 = {self.y1}, x2 = {self.x2}, y2 = {self.y2}, x3 = {self.x3},'
                   f' y3 = {self.y3}. Perimeter = {self.perimeter()}. Color = {self.color}\n')


    def perimeter(self) -> float:
        return float(self.get_side1() + self.get_side2() + self.get_side3())
